src/parser.rs

```diff
@@ -1,7 +1,45 @@
 use std::any::Any;
 
+#[derive(Clone, Debug)]
+pub struct Input<'a> {
+    cursor: usize,
+    value: &'a str,
+}
+
+impl<'a> Input<'a> {
+    pub fn new(input: &'a str) -> Self {
+        Input {
+            cursor: 0,
+            value: input,
+        }
+    }
+
+    pub fn head(&mut self, pattern: &str) -> &'a str {
+        if !self.empty() {
+            let (_, tail) = self.value.split_at(self.cursor);
+            let head = tail.split(pattern).next().unwrap_or("");
+            self.cursor += head.len() + pattern.len();
+            head   
+        } else {
+            ""
+        }
+    }
+
+    pub fn tail(&self) -> &str {
+        if !self.empty() {
+            self.value.split_at(self.cursor).1
+        } else {
+            ""
+        }
+    }
+
+    pub fn empty(&self) -> bool {
+        !(self.cursor < self.value.len())
+    }
+}
+
 pub trait ArgumentChecker<C>: Any {
-    fn satisfies(&self, ctx: &C, input: &str) -> bool;
+    fn satisfies(&self, ctx: &C, input: &mut Input) -> bool;
     /// Returns whether this `ArgumentChecker` will perform
```

Comment:
Can we document what the checker should do to `Input`, i.e. is it expected to advance the cursor?

Reply:
Yes, I'm not quite sure how to handle input probably yet; who is in charge of advancing the input, currently it is the head function.